abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,1471 @@
+upstream.
+
+Metadata (`num_elements` entries) can be retrieved using `proxy_get_map_values`.
+
+Return value (`next_action`) instructs the host environment what action to take:
+- `Continue` means that HTTP/2 METADATA elements should be forwarded downstream.
+- `EndStream` means that HTTP/2 METADATA elements should be forwarded downstream
+  and that the underlying stream should be closed afterwards.
+- `Done` means that HTTP/2 METADATA elements should be forwarded downstream and
+  that this extension shouldn't be called again for this HTTP response.
+- `Pause` instructs the host environment to pause processing until
+  `proxy_resume` is called.
+- `Close` means that the underlying stream should be closed immediately.
+
+### Functions exposed by the host environment
+
+#### `proxy_send_http_response`
+
+* params:
+  - `i32 (uint32_t) response_code`
+  - `i32 (const char*) response_code_details_data`
+  - `i32 (size_t) response_code_details_size`
+  - `i32 (iovec_array_t) response_body_data`
+  - `i32 (size_t) response_body_size`
+  - `i32 (iovec_array_t[]) additional_headers_map_data`
+  - `i32 (size_t) additional_headers_size`
+  - `i32 (uint32_t) grpc_status`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Sends HTTP response to the client, without forwarding request to the upstream.
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for invalid `response_code` and/or `grpc_status`.
+
+Returns `InvalidMemoryAccess` when `response_code_details_data`,
+`response_code_details_size`, `response_body_data`, `response_body_size`,
+`additional_headers_map_data` and/or `additional_headers_size` point to invalid
+memory address.
+
+#### `proxy_resume_http_stream`
+
+* params:
+  - `i32 (proxy_stream_type_t) stream_type`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Resume processing of paused HTTP request or response (`stream_type`) in current
+context.
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for unknown `stream_type`.
+
+Returns `InvalidOperation` when stream wasn't paused.
+
+#### `proxy_close_http_stream`
+
+* params:
+  - `i32 (proxy_stream_type_t) stream_type`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Close HTTP request or response (`stream_type`) in current context.
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for unknown `stream_type`.
+
+Returns `InvalidOperation` when stream is already closed.
+
+
+## Buffers and maps
+
+### Functions exposed by the host environment
+
+#### `proxy_get_buffer`
+
+* params:
+  - `i32 (proxy_buffer_type_t) buffer_type`
+  - `i32 (size_t) offset`
+  - `i32 (size_t) max_size`
+  - `i32 (iovec_array_t*) return_buffer_data`
+  - `i32 (size_t*) return_buffer_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Returns content of the buffer `buffer_type`. At most `max_size` bytes will be
+returned, starting from the `offset`. Returned bytes are written into memory
+slices pointed by a list of vectors (`return_buffer_data`,
+`return_buffer_size`).
```

Comment:
Can you clarify this? The memory is allocated by callout via the allocate API, right? I'm not completely clear on what are in/out params here and how this works. I think more text would help.